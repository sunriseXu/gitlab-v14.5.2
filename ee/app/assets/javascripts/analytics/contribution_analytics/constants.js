import dateFormat from 'dateformat';
import { dateFormats } from '~/analytics/shared/constants';
import { getDateInPast, nMonthsBefore } from '~/lib/utils/datetime_utility';
import { s__ } from '~/locale';

export const CHART_HEIGHT = 350;
export const INNER_CHART_HEIGHT = 200;
export const CHART_X_AXIS_ROTATE = 45;
export const CHART_X_AXIS_NAME_TOP_PADDING = 55;

export const DATE_OPTIONS = [
  {
    text: s__('ContributionAnalytics|Last week'),
    value: dateFormat(getDateInPast(new Date(), 7), dateFormats.isoDate),
  },
  {
    text: s__('ContributionAnalytics|Last month'),
    value: dateFormat(nMonthsBefore(new Date(), 1), dateFormats.isoDate),
  },
  {
    text: s__('ContributionAnalytics|Last 3 months'),
    value: dateFormat(nMonthsBefore(new Date(), 3), dateFormats.isoDate),
  },
];
