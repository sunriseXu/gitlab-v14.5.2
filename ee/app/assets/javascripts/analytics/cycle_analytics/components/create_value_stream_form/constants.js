import { __, s__, sprintf } from '~/locale';

export const NAME_MAX_LENGTH = 100;

export const i18n = {
  FORM_TITLE: s__('CreateValueStreamForm|Create Value Stream'),
  EDIT_FORM_TITLE: s__('CreateValueStreamForm|Edit Value Stream'),
  EDIT_FORM_ACTION: s__('CreateValueStreamForm|Save value stream'),
  FORM_CREATED: s__("CreateValueStreamForm|'%{name}' Value Stream created"),
  FORM_EDITED: s__("CreateValueStreamForm|'%{name}' Value Stream saved"),
  RECOVER_HIDDEN_STAGE: s__('CreateValueStreamForm|Recover hidden stage'),
  RESTORE_HIDDEN_STAGE: s__('CreateValueStreamForm|Restore stage'),
  RESTORE_DEFAULTS: s__('CreateValueStreamForm|Restore defaults'),
  RECOVER_STAGE_TITLE: s__('CreateValueStreamForm|Default stages'),
  RECOVER_STAGES_VISIBLE: s__('CreateValueStreamForm|All default stages are currently visible'),
  SELECT_START_EVENT: s__('CreateValueStreamForm|Select start event'),
  SELECT_END_EVENT: s__('CreateValueStreamForm|Select end event'),
  FORM_FIELD_NAME_LABEL: s__('CreateValueStreamForm|Value Stream name'),
  FORM_FIELD_NAME_PLACEHOLDER: s__('CreateValueStreamForm|Enter value stream name'),
  FORM_FIELD_STAGE_NAME_PLACEHOLDER: s__('CreateValueStreamForm|Enter stage name'),
  FORM_FIELD_START_EVENT: s__('CreateValueStreamForm|Start event'),
  FORM_FIELD_START_EVENT_LABEL: s__('CreateValueStreamForm|Start event label'),
  FORM_FIELD_END_EVENT: s__('CreateValueStreamForm|End event'),
  FORM_FIELD_END_EVENT_LABEL: s__('CreateValueStreamForm|End event label'),
  DEFAULT_FIELD_START_EVENT_LABEL: s__('CreateValueStreamForm|Start event: '),
  DEFAULT_FIELD_END_EVENT_LABEL: s__('CreateValueStreamForm|End event: '),
  BTN_UPDATE_STAGE: s__('CreateValueStreamForm|Update stage'),
  BTN_ADD_STAGE: s__('CreateValueStreamForm|Add stage'),
  BTN_ADD_ANOTHER_STAGE: s__('CreateValueStreamForm|Add another stage'),
  TITLE_EDIT_STAGE: s__('CreateValueStreamForm|Editing stage'),
  TITLE_ADD_STAGE: s__('CreateValueStreamForm|New stage'),
  BTN_CANCEL: __('Cancel'),
  STAGE_INDEX: s__('CreateValueStreamForm|Stage %{index}'),
  HIDDEN_DEFAULT_STAGE: s__('CreateValueStreamForm|%{name} (default)'),
  TEMPLATE_DEFAULT: s__('CreateValueStreamForm|Create from default template'),
  TEMPLATE_BLANK: s__('CreateValueStreamForm|Create from no template'),
  ISSUE_STAGE_END: s__('CreateValueStreamForm|Issue stage end'),
  PLAN_STAGE_START: s__('CreateValueStreamForm|Plan stage start'),
  CODE_STAGE_START: s__('CreateValueStreamForm|Code stage start'),
  DEFAULT_STAGE_FEATURES: s__(
    'CreateValueStreamForm|Default stages can only be hidden or re-ordered',
  ),
};

export const ERRORS = {
  VALUE_STREAM_NAME_MIN_LENGTH: s__('CreateValueStreamForm|Name is required'),
  STAGE_NAME_MIN_LENGTH: s__('CreateValueStreamForm|Stage name is required'),
  MAX_LENGTH: sprintf(s__('CreateValueStreamForm|Maximum length %{maxLength} characters'), {
    maxLength: NAME_MAX_LENGTH,
  }),
  START_EVENT_REQUIRED: s__('CreateValueStreamForm|Please select a start event first'),
  END_EVENT_REQUIRED: s__('CreateValueStreamForm|Please select an end event'),
  STAGE_NAME_EXISTS: s__('CreateValueStreamForm|Stage name already exists'),
  INVALID_EVENT_PAIRS: s__(
    'CreateValueStreamForm|Start event changed, please select a valid end event',
  ),
};

export const STAGE_SORT_DIRECTION = {
  UP: 'UP',
  DOWN: 'DOWN',
};

export const formFieldKeys = [
  'id',
  'name',
  'startEventIdentifier',
  'endEventIdentifier',
  'startEventLabelId',
  'endEventLabelId',
];

export const editableFormFieldKeys = [
  ...formFieldKeys,
  'hidden',
  'description',
  'title',
  'legend',
  'custom',
];

export const defaultFields = formFieldKeys.reduce((acc, field) => ({ ...acc, [field]: null }), {});
export const defaultErrors = formFieldKeys.reduce((acc, field) => ({ ...acc, [field]: [] }), {});

export const defaultCustomStageFields = { ...defaultFields, custom: true };

export const PRESET_OPTIONS_DEFAULT = 'default';
export const PRESET_OPTIONS_BLANK = 'blank';
export const PRESET_OPTIONS = [
  {
    text: i18n.TEMPLATE_DEFAULT,
    value: PRESET_OPTIONS_DEFAULT,
  },
  {
    text: i18n.TEMPLATE_BLANK,
    value: PRESET_OPTIONS_BLANK,
  },
];

// These events can only be set on the back end, they are used in the
// initial configuration of some default stages, but should not be
// selectable by users via the form, they are added here only for display
// purposes when we are editing a default value stream
export const ADDITIONAL_DEFAULT_STAGE_EVENTS = [
  {
    identifier: 'issue_stage_end',
    name: i18n.ISSUE_STAGE_END,
  },
  {
    identifier: 'plan_stage_start',
    name: i18n.PLAN_STAGE_START,
  },
  {
    identifier: 'code_stage_start',
    name: i18n.CODE_STAGE_START,
  },
];
