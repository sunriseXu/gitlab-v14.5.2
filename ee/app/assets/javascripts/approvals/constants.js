import { __, s__ } from '~/locale';

export const TYPE_USER = 'user';
export const TYPE_GROUP = 'group';
export const TYPE_HIDDEN_GROUPS = 'hidden_groups';

export const BRANCH_FETCH_DELAY = 250;
export const ALL_BRANCHES = {
  id: null,
  name: __('All branches'),
};

export const RULE_TYPE_FALLBACK = 'fallback';
export const RULE_TYPE_REGULAR = 'regular';
export const RULE_TYPE_REPORT_APPROVER = 'report_approver';
export const RULE_TYPE_CODE_OWNER = 'code_owner';
export const RULE_TYPE_ANY_APPROVER = 'any_approver';
export const RULE_NAME_ANY_APPROVER = 'All Members';

export const VULNERABILITY_CHECK_NAME = 'Vulnerability-Check';
export const LICENSE_CHECK_NAME = 'License-Check';
export const COVERAGE_CHECK_NAME = 'Coverage-Check';

export const REPORT_TYPE_LICENSE_SCANNING = 'license_scanning';
export const REPORT_TYPE_VULNERABILITY = 'vulnerability';
export const REPORT_TYPE_CODE_COVERAGE = 'code_coverage';

export const APPROVAL_RULE_CONFIGS = {
  [VULNERABILITY_CHECK_NAME]: {
    title: s__('SecurityApprovals|Vulnerability-Check'),
    popoverText: s__(
      'SecurityApprovals|A merge request approval is required when a security report contains a new vulnerability.',
    ),
    documentationText: s__('SecurityApprovals|Learn more about Vulnerability-Check'),
    reportType: REPORT_TYPE_VULNERABILITY,
  },
  [LICENSE_CHECK_NAME]: {
    title: s__('SecurityApprovals|License-Check'),
    popoverText: s__(
      'SecurityApprovals|A merge request approval is required when the license compliance report contains a denied license.',
    ),
    documentationText: s__('SecurityApprovals|Learn more about License-Check'),
    reportType: REPORT_TYPE_LICENSE_SCANNING,
  },
  [COVERAGE_CHECK_NAME]: {
    title: s__('SecurityApprovals|Coverage-Check'),
    popoverText: s__(
      'SecurityApprovals|A merge request approval is required when test coverage declines.',
    ),
    documentationText: s__('SecurityApprovals|Learn more about Coverage-Check'),
    reportType: REPORT_TYPE_CODE_COVERAGE,
  },
};

export const APPROVALS_HELP_PATH = 'user/project/merge_requests/approvals/settings';

export const APPROVAL_SETTINGS_I18N = {
  learnMore: __('Learn more.'),
  approvalSettingsHeader: __('Approval settings'),
  approvalSettingsDescription: __('Define how approval rules are applied to merge requests.'),
  saveChanges: __('Save changes'),
  loadingErrorMessage: s__(
    'ApprovalSettings|There was an error loading merge request approval settings.',
  ),
  savingErrorMessage: s__(
    'ApprovalSettings|There was an error updating merge request approval settings.',
  ),
  savingSuccessMessage: s__('ApprovalSettings|Merge request approval settings have been updated.'),
  lockedByAdmin: s__(
    'ApprovalSettings|This setting is configured at the instance level and can only be changed by an administrator.',
  ),
  lockedByGroupOwner: s__(
    'ApprovalSettings|This setting is configured in %{groupName} and can only be changed by an administrator or group owner.',
  ),
};

export const PROJECT_APPROVAL_SETTINGS_LABELS_I18N = {
  authorApprovalLabel: s__('ApprovalSettings|Prevent approval by author.'),
  preventMrApprovalRuleEditLabel: s__(
    'ApprovalSettings|Prevent editing approval rules in merge requests.',
  ),
  preventCommittersApprovalLabel: s__(
    'ApprovalSettings|Prevent approvals by users who add commits.',
  ),
  requireUserPasswordLabel: s__('ApprovalSettings|Require user password to approve.'),
  removeApprovalsOnPushLabel: s__(
    'ApprovalSettings|Remove all approvals when commits are added to the source branch.',
  ),
};

export const GROUP_APPROVAL_SETTINGS_LABELS_I18N = {
  ...PROJECT_APPROVAL_SETTINGS_LABELS_I18N,
  preventMrApprovalRuleEditLabel: s__(
    'ApprovalSettings|Prevent editing approval rules in projects and merge requests.',
  ),
};

export const APPROVAL_DIALOG_I18N = {
  form: {
    approvalsRequiredLabel: s__('ApprovalRule|Approvals required'),
    approvalTypeLabel: s__('ApprovalRule|Approver Type'),
    approversLabel: s__('ApprovalRule|Add approvers'),
    nameLabel: s__('ApprovalRule|Rule name'),
    nameDescription: s__('ApprovalRule|Examples: QA, Security.'),
    protectedBranchLabel: s__('ApprovalRule|Target branch'),
    protectedBranchDescription: __(
      'Apply this approval rule to all branches or a specific protected branch.',
    ),
    scannersLabel: s__('ApprovalRule|Security scanners'),
    scannersSelectLabel: s__('ApprovalRule|Select scanners'),
    scannersDescription: s__(
      'ApprovalRule|Apply this approval rule to consider only the selected security scanners.',
    ),
    selectAllLabel: s__('ApprovalRule|Select All'),
    allScannersSelectedLabel: s__('ApprovalRule|All scanners'),
    multipleSelectedLabel: s__('ApprovalRule|%{firstLabel} +%{numberOfAdditionalLabels} more'),
    vulnerabilitiesAllowedLabel: s__('ApprovalRule|Vulnerabilities allowed'),
    vulnerabilitiesAllowedDescription: s__(
      'ApprovalRule|Number of vulnerabilities allowed before approval rule is triggered.',
    ),
    vulnerabilityStatesLabel: s__('ApprovalRule|Vulnerability states'),
    vulnerabilityStatesDescription: s__(
      'ApprovalRule|Apply this approval rule to consider only the selected vulnerability states.',
    ),
    vulnerabilityStatesSelectLabel: s__('ApprovalRule|Select vulnerability states'),
    allVulnerabilityStatesSelectedLabel: s__('ApprovalRule|All vulnerability states'),
    severityLevelsLabel: s__('ApprovalRule|Severity levels'),
    severityLevelsDescription: s__(
      'ApprovalRule|Apply this approval rule to consider only the selected severity levels.',
    ),
    severityLevelsSelectLabel: s__('ApprovalRule|Select severity levels'),
    allSeverityLevelsSelectedLabel: s__('ApprovalRule|All severity levels'),
  },
  validations: {
    approvalsRequiredNegativeNumber: __('Please enter a non-negative number'),
    approvalsRequiredNotNumber: __('Please enter a valid number'),
    approvalsRequiredMinimum: __(
      'Please enter a number greater than %{number} (from the project settings)',
    ),
    approversRequired: __('Please select and add a member'),
    branchesRequired: __('Please select a valid target branch'),
    ruleNameTaken: __('Rule name is already taken.'),
    ruleNameMissing: __('Please provide a name'),
    scannersRequired: s__('ApprovalRule|Please select at least one security scanner'),
    vulnerabilitiesAllowedMinimum: s__(
      'ApprovalRule|Please enter a number equal or greater than zero',
    ),
    severityLevelsRequired: s__('ApprovalRule|Please select at least one severity level'),
    vulnerabilityStatesRequired: s__('ApprovalRule|Please select at least one vulnerability state'),
  },
};

export const APPROVAL_VULNERABILITY_STATES = {
  newly_detected: s__('ApprovalRule|Newly detected'),
  detected: s__('ApprovalRule|Previously detected'),
  confirmed: s__('ApprovalRule|Confirmed'),
  dismissed: s__('ApprovalRule|Dismissed'),
  resolved: s__('ApprovalRule|Resolved'),
};
