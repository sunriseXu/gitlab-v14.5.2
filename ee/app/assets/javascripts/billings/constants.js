export const TABLE_TYPE_DEFAULT = 'default';
export const TABLE_TYPE_FREE = 'free';
export const TABLE_TYPE_TRIAL = 'trial';

export const DAYS_FOR_RENEWAL = 15;
