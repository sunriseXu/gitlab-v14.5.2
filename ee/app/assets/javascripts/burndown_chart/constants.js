export { Namespace } from '../iterations/constants';

export const Unit = {
  count: 'count',
  weight: 'weight',
};
