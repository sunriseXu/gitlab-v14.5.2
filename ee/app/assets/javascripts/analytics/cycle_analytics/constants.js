import { getGroupValueStreamMetrics } from 'ee/api/analytics_api';
import { METRIC_TYPE_SUMMARY, METRIC_TYPE_TIME_SUMMARY } from '~/api/analytics_api';
import { OVERVIEW_STAGE_ID } from '~/cycle_analytics/constants';
import { __ } from '~/locale';

export const EVENTS_LIST_ITEM_LIMIT = 50;

export const TASKS_BY_TYPE_SUBJECT_ISSUE = 'Issue';
export const TASKS_BY_TYPE_SUBJECT_MERGE_REQUEST = 'MergeRequest';
export const TASKS_BY_TYPE_MAX_LABELS = 15;

export const TASKS_BY_TYPE_SUBJECT_FILTER_OPTIONS = {
  [TASKS_BY_TYPE_SUBJECT_ISSUE]: __('Issues'),
  [TASKS_BY_TYPE_SUBJECT_MERGE_REQUEST]: __('Merge Requests'),
};

export const TASKS_BY_TYPE_FILTERS = {
  SUBJECT: 'SUBJECT',
  LABEL: 'LABEL',
};

export const DEFAULT_VALUE_STREAM_ID = 'default';

export const FETCH_VALUE_STREAM_DATA = 'fetchValueStreamData';

export const OVERVIEW_STAGE_CONFIG = {
  id: OVERVIEW_STAGE_ID,
  title: __('Overview'),
  icon: 'home',
};

export const METRICS_REQUESTS = [
  {
    endpoint: METRIC_TYPE_TIME_SUMMARY,
    request: getGroupValueStreamMetrics,
    name: __('time summary'),
  },
  {
    endpoint: METRIC_TYPE_SUMMARY,
    request: getGroupValueStreamMetrics,
    name: __('recent activity'),
  },
];
