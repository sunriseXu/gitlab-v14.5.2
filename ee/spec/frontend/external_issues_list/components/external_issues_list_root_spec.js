import { GlAlert } from '@gitlab/ui';
import * as Sentry from '@sentry/browser';
import { shallowMount, createLocalVue, mount } from '@vue/test-utils';
import MockAdapter from 'axios-mock-adapter';
import VueApollo from 'vue-apollo';

import ExternalIssuesListRoot from 'ee/external_issues_list/components/external_issues_list_root.vue';
import jiraIssuesResolver from 'ee/integrations/jira/issues_list/graphql/resolvers/jira_issues';

import createMockApollo from 'helpers/mock_apollo_helper';
import waitForPromises from 'helpers/wait_for_promises';

import IssuableList from '~/issuable_list/components/issuable_list_root.vue';
import { i18n } from '~/issues_list/constants';
import axios from '~/lib/utils/axios_utils';
import httpStatus from '~/lib/utils/http_status';

import {
  mockProvide,
  mockJiraIssues as mockExternalIssues,
  mockJiraIssue4 as mockJiraIssueNoReference,
} from '../mock_data';

jest.mock('~/flash');
jest.mock('~/issuable_list/constants', () => ({
  DEFAULT_PAGE_SIZE: 2,
  IssuableStates: jest.requireActual('~/issuable_list/constants').IssuableStates,
  IssuableListTabs: jest.requireActual('~/issuable_list/constants').IssuableListTabs,
  AvailableSortOptions: jest.requireActual('~/issuable_list/constants').AvailableSortOptions,
}));
jest.mock(
  '~/vue_shared/components/filtered_search_bar/tokens/label_token.vue',
  () => 'LabelTokenMock',
);

const resolvedValue = {
  headers: {
    'x-page': 1,
    'x-total': mockExternalIssues.length,
  },
  data: mockExternalIssues,
};

const localVue = createLocalVue();

const resolvers = {
  Query: {
    externalIssues: jiraIssuesResolver,
  },
};

function createMockApolloProvider(mockResolvers = resolvers) {
  localVue.use(VueApollo);
  return createMockApollo([], mockResolvers);
}

describe('ExternalIssuesListRoot', () => {
  let wrapper;
  let mock;

  const mockSearchTerm = 'test issue';
  const mockLabel = 'ecosystem';

  const findIssuableList = () => wrapper.findComponent(IssuableList);
  const findAlert = () => wrapper.findComponent(GlAlert);
  const findAlertMessage = () => findAlert().find('span');
  const createLabelFilterEvent = (data) => ({ type: 'labels', value: { data } });
  const createSearchFilterEvent = (data) => ({ type: 'filtered-search-term', value: { data } });

  const expectErrorHandling = (expectedRenderedErrorMessage) => {
    const issuesList = findIssuableList();
    const alert = findAlert();

    expect(issuesList.exists()).toBe(false);

    expect(alert.exists()).toBe(true);
    expect(alert.text()).toBe(expectedRenderedErrorMessage);
    expect(Sentry.captureException).toHaveBeenCalledWith(expect.any(Error));
  };

  const createComponent = ({
    apolloProvider = createMockApolloProvider(),
    provide = mockProvide,
    initialFilterParams = {},
  } = {}) => {
    wrapper = shallowMount(ExternalIssuesListRoot, {
      propsData: {
        initialFilterParams,
      },
      provide,
      localVue,
      apolloProvider,
    });
  };

  beforeEach(() => {
    mock = new MockAdapter(axios);
  });

  afterEach(() => {
    wrapper.destroy();
    mock.restore();
  });

  describe('while loading', () => {
    it('sets issuesListLoading to `true`', async () => {
      jest.spyOn(axios, 'get').mockResolvedValue(new Promise(() => {}));

      createComponent();
      await wrapper.vm.$nextTick();

      const issuableList = findIssuableList();
      expect(issuableList.props('issuablesLoading')).toBe(true);
    });

    it('calls `axios.get` with `issuesFetchPath` and query params', async () => {
      jest.spyOn(axios, 'get');

      createComponent();
      await waitForPromises();

      expect(axios.get).toHaveBeenCalledWith(
        mockProvide.issuesFetchPath,
        expect.objectContaining({
          params: {
            with_labels_details: true,
            page: wrapper.vm.currentPage,
            per_page: wrapper.vm.$options.defaultPageSize,
            state: wrapper.vm.currentState,
            sort: wrapper.vm.sortedBy,
            search: wrapper.vm.filterParams.search,
          },
        }),
      );
    });
  });

  describe('with `initialFilterParams` prop', () => {
    beforeEach(async () => {
      jest.spyOn(axios, 'get').mockResolvedValue(resolvedValue);

      createComponent({
        initialFilterParams: {
          labels: [mockLabel],
          search: mockSearchTerm,
        },
      });
      await waitForPromises();
    });

    it('renders issuable-list component with correct props', () => {
      const issuableList = findIssuableList();

      expect(issuableList.props('initialFilterValue')).toEqual([
        { type: 'labels', value: { data: mockLabel } },
        { type: 'filtered-search-term', value: { data: mockSearchTerm } },
      ]);
      expect(issuableList.props('urlParams').search).toBe(mockSearchTerm);
    });
  });

  describe('when request succeeds', () => {
    beforeEach(async () => {
      jest.spyOn(axios, 'get').mockResolvedValue(resolvedValue);

      createComponent();
      await waitForPromises();
    });

    it('renders issuable-list component with correct props', () => {
      const issuableList = findIssuableList();
      expect(issuableList.exists()).toBe(true);
      expect(issuableList.props()).toMatchSnapshot();
    });

    describe('issuable-list reference section', () => {
      it('renders issuable-list component with correct reference', async () => {
        jest.spyOn(axios, 'get').mockResolvedValue(resolvedValue);

        wrapper = mount(ExternalIssuesListRoot, {
          propsData: {
            initialFilterParams: {},
          },
          provide: mockProvide,
          localVue,
          apolloProvider: createMockApolloProvider(),
        });
        await waitForPromises();
        expect(wrapper.find('.issuable-info').text()).toContain(
          resolvedValue.data[0].references.relative,
        );
      });

      it('renders issuable-list component with id when references is not presence', async () => {
        jest.spyOn(axios, 'get').mockResolvedValue({
          ...resolvedValue,
          data: [mockJiraIssueNoReference],
        });

        wrapper = mount(ExternalIssuesListRoot, {
          propsData: {
            initialFilterParams: {},
          },
          provide: mockProvide,
          localVue,
          apolloProvider: createMockApolloProvider(),
        });
        await waitForPromises();
        // Since Jira transformer transforms references.relative into id, we can only test
        // whether it exists.
        expect(wrapper.find('.issuable-info').exists()).toBe(false);
      });
    });

    describe('issuable-list events', () => {
      it('"click-tab" event executes GET request correctly', async () => {
        const issuableList = findIssuableList();

        issuableList.vm.$emit('click-tab', 'closed');
        await waitForPromises();

        expect(axios.get).toHaveBeenCalledWith(mockProvide.issuesFetchPath, {
          params: {
            labels: undefined,
            page: 1,
            per_page: 2,
            search: undefined,
            sort: 'created_desc',
            state: 'closed',
            with_labels_details: true,
          },
        });
        expect(issuableList.props('currentTab')).toBe('closed');
      });

      it('"page-change" event executes GET request correctly', async () => {
        const mockPage = 2;
        const issuableList = findIssuableList();
        jest.spyOn(axios, 'get').mockResolvedValue({
          ...resolvedValue,
          headers: { 'x-page': mockPage, 'x-total': mockExternalIssues.length },
        });

        issuableList.vm.$emit('page-change', mockPage);
        await waitForPromises();

        expect(axios.get).toHaveBeenCalledWith(mockProvide.issuesFetchPath, {
          params: {
            labels: undefined,
            page: mockPage,
            per_page: 2,
            search: undefined,
            sort: 'created_desc',
            state: 'opened',
            with_labels_details: true,
          },
        });

        await wrapper.vm.$nextTick();
        expect(issuableList.props()).toMatchObject({
          currentPage: mockPage,
          previousPage: mockPage - 1,
          nextPage: mockPage + 1,
        });
      });

      it('"sort" event executes GET request correctly', async () => {
        const mockSortBy = 'updated_asc';
        const issuableList = findIssuableList();

        issuableList.vm.$emit('sort', mockSortBy);
        await waitForPromises();

        expect(axios.get).toHaveBeenCalledWith(mockProvide.issuesFetchPath, {
          params: {
            labels: undefined,
            page: 1,
            per_page: 2,
            search: undefined,
            sort: 'created_desc',
            state: 'opened',
            with_labels_details: true,
          },
        });
        expect(issuableList.props('initialSortBy')).toBe(mockSortBy);
      });

      it.each`
        desc                        | input                                                                           | expected
        ${'with label and search'}  | ${[createLabelFilterEvent(mockLabel), createSearchFilterEvent(mockSearchTerm)]} | ${{ labels: [mockLabel], search: mockSearchTerm }}
        ${'with multiple lables'}   | ${[createLabelFilterEvent('label1'), createLabelFilterEvent('label2')]}         | ${{ labels: ['label1', 'label2'], search: undefined }}
        ${'with multiple searches'} | ${[createSearchFilterEvent('foo bar'), createSearchFilterEvent('lorem')]}       | ${{ labels: undefined, search: 'foo bar lorem' }}
      `(
        '$desc, filter event sets "filterParams" value and calls fetchIssues',
        async ({ input, expected }) => {
          const issuableList = findIssuableList();

          issuableList.vm.$emit('filter', input);
          await waitForPromises();

          expect(axios.get).toHaveBeenCalledWith(mockProvide.issuesFetchPath, {
            params: {
              page: 1,
              per_page: 2,
              sort: 'created_desc',
              state: 'opened',
              with_labels_details: true,
              ...expected,
            },
          });
        },
      );
    });
  });

  describe('error handling', () => {
    beforeEach(() => {
      jest.spyOn(Sentry, 'captureException');
    });

    describe('when request fails', () => {
      it.each`
        APIErrors                                         | expectedRenderedErrorText
        ${['API error']}                                  | ${'API error'}
        ${['API <a href="gitlab.com">error</a>']}         | ${'API error'}
        ${['API <script src="hax0r.xyz">error</script>']} | ${'API'}
        ${undefined}                                      | ${i18n.errorFetchingIssues}
      `(
        'displays error alert with "$expectedRenderedErrorText" when API responds with "$APIErrors"',
        async ({ APIErrors, expectedRenderedErrorText }) => {
          jest.spyOn(axios, 'get');
          mock
            .onGet(mockProvide.issuesFetchPath)
            .replyOnce(httpStatus.INTERNAL_SERVER_ERROR, { errors: APIErrors });

          createComponent();
          await waitForPromises();

          expectErrorHandling(expectedRenderedErrorText);
          expect(findAlertMessage().html()).toMatchSnapshot();
        },
      );
    });

    describe('when GraphQL network error is encountered', () => {
      it('displays error alert with default error message', async () => {
        createComponent({
          apolloProvider: createMockApolloProvider({
            Query: {
              externalIssues: jest.fn().mockRejectedValue(new Error('GraphQL networkError')),
            },
          }),
        });
        await waitForPromises();

        expectErrorHandling(i18n.errorFetchingIssues);
      });
    });
  });

  describe('pagination', () => {
    it.each`
      scenario                 | issues                | shouldShowPaginationControls
      ${'returns no issues'}   | ${[]}                 | ${false}
      ${`returns some issues`} | ${mockExternalIssues} | ${true}
    `(
      'sets `showPaginationControls` prop to $shouldShowPaginationControls when request $scenario',
      async ({ issues, shouldShowPaginationControls }) => {
        jest.spyOn(axios, 'get');
        mock.onGet(mockProvide.issuesFetchPath).replyOnce(httpStatus.OK, issues, {
          'x-page': 1,
          'x-total': issues.length,
        });

        createComponent();
        await waitForPromises();

        expect(findIssuableList().props('showPaginationControls')).toBe(
          shouldShowPaginationControls,
        );
      },
    );
  });
});
