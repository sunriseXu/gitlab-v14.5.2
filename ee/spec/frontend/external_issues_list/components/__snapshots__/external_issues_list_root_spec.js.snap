// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExternalIssuesListRoot error handling when request fails displays error alert with "API error" when API responds with "["API <a href=\\"gitlab.com\\">error</a>"]" 1`] = `"<span>API <a href=\\"gitlab.com\\">error</a></span>"`;

exports[`ExternalIssuesListRoot error handling when request fails displays error alert with "API error" when API responds with "["API error"]" 1`] = `"<span>API error</span>"`;

exports[`ExternalIssuesListRoot error handling when request fails displays error alert with "API" when API responds with "["API <script src=\\"hax0r.xyz\\">error</script>"]" 1`] = `"<span>API </span>"`;

exports[`ExternalIssuesListRoot error handling when request fails displays error alert with "An error occurred while loading issues" when API responds with "undefined" 1`] = `"<span>An error occurred while loading issues</span>"`;

exports[`ExternalIssuesListRoot when request succeeds renders issuable-list component with correct props 1`] = `
Object {
  "currentPage": 1,
  "currentTab": "opened",
  "defaultPageSize": 2,
  "enableLabelPermalinks": true,
  "hasNextPage": false,
  "hasPreviousPage": false,
  "initialFilterValue": Array [],
  "initialSortBy": "created_desc",
  "isManualOrdering": false,
  "issuableSymbol": "#",
  "issuables": Array [
    Object {
      "assignees": Array [
        Object {
          "avatarUrl": null,
          "name": "Kushal Pandya",
          "webUrl": "https://gitlab-jira.atlassian.net/people/1920938475",
        },
      ],
      "author": Object {
        "avatarUrl": null,
        "name": "jhope",
        "webUrl": "https://gitlab-jira.atlassian.net/people/5e32f803e127810e82875bc1",
      },
      "closedAt": null,
      "createdAt": "2020-03-19T14:31:51.281Z",
      "externalTracker": "jira",
      "gitlabWebUrl": "",
      "id": 31596,
      "labels": Array [
        Object {
          "color": "#0052CC",
          "name": "backend",
          "textColor": "#FFFFFF",
          "title": "backend",
        },
      ],
      "projectId": 1,
      "references": Object {
        "relative": "IG-31596",
      },
      "status": "Selected for Development",
      "title": "Eius fuga voluptates.",
      "updatedAt": "2020-10-20T07:01:45.865Z",
      "webUrl": "https://gitlab-jira.atlassian.net/browse/IG-31596",
    },
    Object {
      "assignees": Array [],
      "author": Object {
        "avatarUrl": null,
        "name": "Gabe Weaver",
        "webUrl": "https://gitlab-jira.atlassian.net/people/5e320a31fe03e20c9d1dccde",
      },
      "closedAt": null,
      "createdAt": "2020-03-19T14:31:50.677Z",
      "externalTracker": "jira",
      "gitlabWebUrl": "",
      "id": 31595,
      "labels": Array [],
      "projectId": 1,
      "references": Object {
        "relative": "IG-31595",
      },
      "status": "Backlog",
      "title": "Hic sit sint ducimus ea et sint.",
      "updatedAt": "2020-03-19T14:31:50.677Z",
      "webUrl": "https://gitlab-jira.atlassian.net/browse/IG-31595",
    },
    Object {
      "assignees": Array [],
      "author": Object {
        "avatarUrl": null,
        "name": "Gabe Weaver",
        "webUrl": "https://gitlab-jira.atlassian.net/people/5e320a31fe03e20c9d1dccde",
      },
      "closedAt": null,
      "createdAt": "2020-03-19T14:31:50.012Z",
      "externalTracker": "jira",
      "gitlabWebUrl": "",
      "id": 31594,
      "labels": Array [],
      "projectId": 1,
      "references": Object {
        "relative": "IG-31594",
      },
      "status": "Backlog",
      "title": "Alias ut modi est labore.",
      "updatedAt": "2020-03-19T14:31:50.012Z",
      "webUrl": "https://gitlab-jira.atlassian.net/browse/IG-31594",
    },
  ],
  "issuablesLoading": false,
  "labelFilterParam": "labels",
  "namespace": "gitlab-org/gitlab-test",
  "nextPage": 2,
  "previousPage": 0,
  "recentSearchesStorageKey": "jira_issues",
  "searchInputPlaceholder": "Search Jira issues",
  "searchTokens": Array [
    Object {
      "defaultLabels": Array [],
      "fetchLabels": [Function],
      "icon": "labels",
      "operators": Array [
        Object {
          "description": "is",
          "value": "=",
        },
      ],
      "symbol": "~",
      "title": "Label",
      "token": "LabelTokenMock",
      "type": "labels",
      "unique": false,
    },
  ],
  "showBulkEditSidebar": false,
  "showPaginationControls": true,
  "sortOptions": Array [
    Object {
      "id": 1,
      "sortDirection": Object {
        "ascending": "created_asc",
        "descending": "created_desc",
      },
      "title": "Created date",
    },
    Object {
      "id": 2,
      "sortDirection": Object {
        "ascending": "updated_asc",
        "descending": "updated_desc",
      },
      "title": "Last updated",
    },
  ],
  "tabCounts": null,
  "tabs": Array [
    Object {
      "id": "state-opened",
      "name": "opened",
      "title": "Open",
      "titleTooltip": "Filter by issues that are currently opened.",
    },
    Object {
      "id": "state-closed",
      "name": "closed",
      "title": "Closed",
      "titleTooltip": "Filter by issues that are currently closed.",
    },
    Object {
      "id": "state-all",
      "name": "all",
      "title": "All",
      "titleTooltip": "Show all issues.",
    },
  ],
  "totalItems": 3,
  "urlParams": Object {
    "labels[]": undefined,
    "search": undefined,
  },
  "useKeysetPagination": false,
}
`;
